chain/epoch-manager/src/adapter.rs

```diff
@@ -106,9 +107,34 @@ pub trait EpochManagerAdapter: Send + Sync {
     /// Get [`EpochId`] from a block belonging to the epoch.
     fn get_epoch_id(&self, block_hash: &CryptoHash) -> Result<EpochId, Error>;
 
+    /// Which of the two epochs is earlier.
+    ///
+    /// This is well-defined because finality gadget guarantees that we cannot
+    /// have two different epochs on two forks.
+    fn compare_epoch_id(
+        &self,
+        epoch_id: &EpochId,
+        other_epoch_id: &EpochId,
+    ) -> Result<Ordering, Error>;
+
     /// Get epoch start from a block belonging to the epoch.
     fn get_epoch_start_height(&self, block_hash: &CryptoHash) -> Result<BlockHeight, Error>;
 
+    /// Get previous epoch id by hash of previous block.
+    fn get_prev_epoch_id_from_prev_block(
+        &self,
+        prev_block_hash: &CryptoHash,
+    ) -> Result<EpochId, Error>;
+
+    /// _If_ the next epoch will use a new protocol version, returns an
+    /// estimated block height for when the epoch switch occurs.
+    ///
+    /// This is very approximate and is used for logging only.
```

Comment:
nit: should we change the name to reflect this ?

get_approximate_protocol_upgrade_block_height ?